### ISSUE TITLE
Requesting a list of Device twin reportedStateCallback() status_code values and meanings

### ISSUE THREAD

**Issue by liudr** [Azure/azure-iot-sdk-c#1351]
[update]
I am requesting a list of status_code and their meanings returned by the following function reportedStateCallback(). My port is esp-azure but this status_code should be generic. Thanks.

[original]
I'm using the esp-azure port but the c port has the same function, a callback function when device completes sending device twin update:

```
static void reportedStateCallback(int status_code, void* userContextCallback)
{
    (void)userContextCallback;
    printf("Device Twin reported properties update completed with result: %d\r\n", status_code);
}

```
I wonder where to find explanations to the returned status_code values. I have seen these two values:
240, 400
What does the status_code mean?

https://github.com/Azure/azure-iot-sdk-c/blob/704f9a9bfaa820829689bbb4c1ebc8146b488c7b/serializer/samples/devicetwin_simplesample/devicetwin_simplesample.c#L79

**Comment by ewertons** [comment]
Hi @liudr ,

we apologize but looks like we do not have that documented. 
To unblock you, the status_code will be 204 if the reported properties were updated successfully, otherwise status_code will be 400.

**Comment by liudr** [comment]
Thanks @ewertons for clarifying. Any way to find these values somewhere in the sdk? I found in iothubtransport_amqp_twin_messenger.h a bunch of definitions that might be relevant but those defs and macros prevent me from knowing the actual values. Any help is appreciated.

Running the device twin sample code, I never got this return value printed. I suppose the callback was never called:

```
static void reportedStateCallback(int status_code, void* userContextCallback)
{
    (void)userContextCallback;
    printf("Device Twin reported properties update completed with result: %d\r\n", status_code);
}
```
I think this is worse than the callback returning 400. Any idea why it failed to run? Again I'm using thes esp-azure port, not this port but both are c ports.

My device twin is over 1300 bytes. With my shorter device twins at 944 bytes and ever at 1200 bytes the callback runs and produces 204.

I thus suspect the esp-azure library doesn't allocate enough buffer to actually get all bytes and runs into problem when the message is longer than expected. I raised a separate issue here in their repo [#72](https://github.com/espressif/esp-azure/issues/72)

Hope this gets resolved.

**Comment by ewertons** [comment]
@liudr ,
indeed it might be an issue that goes beyond the Azure IoT SDK capability.

Running this sample on a regular Linux works as expected.

```bash
ewertons@ubuntu:~/azure-iot-sdk-c/cmake/serializer/samples/devicetwin_simplesample$ ./devicetwin_simplesample
Reported state will be send to IoTHub
press ENTER to end the sample
-> Header (AMQP 0.1.0.0)
<- Header (AMQP 0.1.0.0)
...
IoTHub: reported properties delivered with status_code = 200
```

**Comment by ewertons** [comment]
In this case, would it be reasonable to follow up on the issue filed against the espressif repo, and close this issue?

**Comment by liudr** [comment]
@ewertons 
Thanks for running a test. I think it is still appropriate regarding the meaning of status_code that other developers might be also looking for answers. Such as your test returning 200, mine 204, and 400. In best-case scenario everyone gets 200 I guess. But many of us need to plan for less-than-best-case scenario so we still need a full list of status_code and their meanings.

I just modified my OP to reflect the request for status_code and their meanings.

Best,

**Comment by ewertons** [comment]
It makes sense. Thanks @liudr

**Comment by ericwol-msft** [comment]
These error code are generated on the IoT Hub service side. 

https://github.com/Azure/azure-sdk-for-c/blob/main/sdk/inc/azure/iot/az_iot_common.h#L74
https://github.com/Azure/azure-sdk-for-c/blob/main/sdk/inc/azure/iot/az_iot_common.h#L261
https://github.com/Azure/azure-sdk-for-c/blob/main/sdk/inc/azure/iot/az_iot_common.h#L273
